# Food Delivery Orders & ETA Logistics — Exploratory Data Analysis

## Objective

The objective of this project is to perform Exploratory Data Analysis
on food delivery order data to understand:

- Delivery performance
- Customer ordering patterns
- Restaurant performance
- Delivery time
- Traffic and weather impact
- Peak ordering periods
- Ratings
- Operational factors affecting delivery

In [8]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns


In [9]:
# Import Csv File 
df = pd.read_csv(r"C:\Users\rampa\DS Python Class\Food-Delivery-EDA\Data\Row Data\food_delivery_orders_dataset.csv")

## Infromastion abount The data set 

In [10]:
df.head()

,order_id,customer_id,restaurant_id,driver_id,order_timestamp,order_date,order_hour,day_of_week,is_weekend,city,...,traffic_level,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,order_status,cancellation_reason,customer_rating
0,ORD_000001,CUS_00001,RES_0008,DRV_0075,2025-10-05 15:54:00,2025-10-05,15,Sunday,1,City_C,...,Medium,14,4.4,25,25,29.0,0.0,Completed,NaN,4.6
1,ORD_000002,CUS_00013,RES_0024,DRV_0131,2025-10-01 20:00:00,2025-10-01,20,Wednesday,0,City_A,...,High,44,4.7,12,20,24.0,0.0,Completed,NaN,5.0
2,ORD_000003,CUS_00049,RES_0013,DRV_0087,2025-10-07 18:24:00,2025-10-07,18,Tuesday,0,City_C,...,High,1,5.0,11,63,72.0,1.0,Completed,NaN,3.6
3,ORD_000004,CUS_00007,RES_0001,DRV_0177,2025-10-30 16:54:00,2025-10-30,16,Thursday,0,City_A,...,Low,50,4.3,12,23,31.0,1.0,Completed,NaN,4.1
4,ORD_000005,CUS_00325,RES_0029,DRV_0114,2025-10-15 18:34:00,2025-10-15,18,Wednesday,0,City_B,...,Medium,21,4.0,9,32,42.0,1.0,Completed,NaN,3.9


In [11]:
df.tail()

,order_id,customer_id,restaurant_id,driver_id,order_timestamp,order_date,order_hour,day_of_week,is_weekend,city,...,traffic_level,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,order_status,cancellation_reason,customer_rating
49995,ORD_049996,CUS_00280,RES_0130,DRV_0112,2025-10-25 12:16:00,2025-10-25,12,Saturday,1,City_B,...,Medium,34,5.0,11,19,21.0,0.0,Completed,NaN,4.4
49996,ORD_049997,CUS_00893,RES_0028,DRV_0036,2025-10-05 19:48:00,2025-10-05,19,Sunday,1,City_B,...,High,49,5.0,9,28,29.0,0.0,Completed,NaN,4.8
49997,ORD_049998,CUS_01618,RES_0001,DRV_0071,2025-10-08 12:18:00,2025-10-08,12,Wednesday,0,City_A,...,High,54,4.1,11,18,16.0,0.0,Completed,NaN,4.7
49998,ORD_049999,CUS_00025,RES_0057,DRV_0103,2025-10-20 17:39:00,2025-10-20,17,Monday,0,City_A,...,High,33,4.0,20,17,27.0,1.0,Completed,NaN,3.8
49999,ORD_050000,CUS_00038,RES_0139,DRV_0131,2025-10-13 17:36:00,2025-10-13,17,Monday,0,City_A,...,Low,44,4.7,12,30,40.0,1.0,Completed,NaN,4.5


## Basic Dataset Understanding

In [12]:
df.shape

(50000, 37)

In [13]:
df.columns.tolist()

['order_id',
 'customer_id',
 'restaurant_id',
 'driver_id',
 'order_timestamp',
 'order_date',
 'order_hour',
 'day_of_week',
 'is_weekend',
 'city',
 'delivery_area',
 'customer_age',
 'customer_type',
 'restaurant_type',
 'restaurant_primary_category',
 'restaurant_rating',
 'items_count',
 'subtotal',
 'discount_percent',
 'tax_amount',
 'service_fee',
 'delivery_fee',
 'order_total',
 'payment_method',
 'tip_amount',
 'distance_km',
 'weather',
 'traffic_level',
 'delivery_partner_experience_months',
 'delivery_partner_rating',
 'restaurant_preparation_time_minutes',
 'estimated_delivery_time_minutes',
 'actual_delivery_time_minutes',
 'late_delivery',
 'order_status',
 'cancellation_reason',
 'customer_rating']

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 37 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   order_id                             50000 non-null  object 
 1   customer_id                          50000 non-null  object 
 2   restaurant_id                        50000 non-null  object 
 3   driver_id                            50000 non-null  object 
 4   order_timestamp                      50000 non-null  object 
 5   order_date                           50000 non-null  object 
 6   order_hour                           50000 non-null  int64  
 7   day_of_week                          50000 non-null  object 
 8   is_weekend                           50000 non-null  int64  
 9   city                                 50000 non-null  object 
 10  delivery_area                        50000 non-null  object 
 11  customer_age                

In [15]:
df.describe()

,order_hour,is_weekend,customer_age,restaurant_rating,items_count,subtotal,discount_percent,tax_amount,service_fee,delivery_fee,order_total,tip_amount,distance_km,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,customer_rating
count,50000.000000,50000.00000,50000.000000,50000.000000,50000.00000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,49132.000000,50000.000000,50000.0000,50000.000000,50000.000000,50000.000000,49132.000000,49132.000000,49132.000000
mean,14.465900,0.26712,27.686580,4.207956,2.91002,30.897938,4.495700,4.128796,1.474600,3.075108,38.169501,1.610076,4.730632,29.7332,4.495062,15.549600,30.950580,35.152365,0.365709,4.295946
std,5.231155,0.44246,7.537796,0.450977,2.02907,27.068965,6.430008,3.636275,1.378101,2.299115,31.054669,2.878464,2.780190,17.4052,0.349198,7.360099,12.825519,14.701154,0.481633,0.530947
min,0.000000,0.00000,18.000000,2.800000,1.00000,5.100000,0.000000,0.570000,0.130000,0.000000,4.870000,0.000000,0.800000,1.0000,3.200000,5.000000,8.000000,10.000000,0.000000,1.800000
25%,11.000000,0.00000,21.000000,3.900000,1.00000,13.100000,0.000000,1.730000,0.580000,0.000000,17.430000,0.000000,2.800000,14.0000,4.300000,10.000000,22.000000,25.000000,0.000000,4.000000
50%,14.000000,0.00000,27.000000,4.200000,2.00000,22.825000,0.000000,2.890000,1.040000,3.500000,28.350000,0.000000,4.100000,29.0000,4.500000,14.000000,28.000000,33.000000,0.000000,4.400000
75%,19.000000,1.00000,33.000000,4.600000,4.00000,40.120000,10.000000,5.390000,1.890000,4.500000,48.660000,2.280000,5.900000,44.0000,4.800000,19.000000,37.000000,42.000000,1.000000,4.700000
max,23.000000,1.00000,55.000000,5.000000,10.00000,362.980000,20.000000,50.820000,25.210000,12.000000,443.210000,53.650000,20.000000,60.0000,5.000000,73.000000,128.000000,172.000000,1.000000,5.000000


In [16]:
df.describe()

,order_hour,is_weekend,customer_age,restaurant_rating,items_count,subtotal,discount_percent,tax_amount,service_fee,delivery_fee,order_total,tip_amount,distance_km,delivery_partner_experience_months,delivery_partner_rating,restaurant_preparation_time_minutes,estimated_delivery_time_minutes,actual_delivery_time_minutes,late_delivery,customer_rating
count,50000.000000,50000.00000,50000.000000,50000.000000,50000.00000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,49132.000000,50000.000000,50000.0000,50000.000000,50000.000000,50000.000000,49132.000000,49132.000000,49132.000000
mean,14.465900,0.26712,27.686580,4.207956,2.91002,30.897938,4.495700,4.128796,1.474600,3.075108,38.169501,1.610076,4.730632,29.7332,4.495062,15.549600,30.950580,35.152365,0.365709,4.295946
std,5.231155,0.44246,7.537796,0.450977,2.02907,27.068965,6.430008,3.636275,1.378101,2.299115,31.054669,2.878464,2.780190,17.4052,0.349198,7.360099,12.825519,14.701154,0.481633,0.530947
min,0.000000,0.00000,18.000000,2.800000,1.00000,5.100000,0.000000,0.570000,0.130000,0.000000,4.870000,0.000000,0.800000,1.0000,3.200000,5.000000,8.000000,10.000000,0.000000,1.800000
25%,11.000000,0.00000,21.000000,3.900000,1.00000,13.100000,0.000000,1.730000,0.580000,0.000000,17.430000,0.000000,2.800000,14.0000,4.300000,10.000000,22.000000,25.000000,0.000000,4.000000
50%,14.000000,0.00000,27.000000,4.200000,2.00000,22.825000,0.000000,2.890000,1.040000,3.500000,28.350000,0.000000,4.100000,29.0000,4.500000,14.000000,28.000000,33.000000,0.000000,4.400000
75%,19.000000,1.00000,33.000000,4.600000,4.00000,40.120000,10.000000,5.390000,1.890000,4.500000,48.660000,2.280000,5.900000,44.0000,4.800000,19.000000,37.000000,42.000000,1.000000,4.700000
max,23.000000,1.00000,55.000000,5.000000,10.00000,362.980000,20.000000,50.820000,25.210000,12.000000,443.210000,53.650000,20.000000,60.0000,5.000000,73.000000,128.000000,172.000000,1.000000,5.000000


In [17]:
df.nunique().sort_values()

is_weekend                                 2
order_status                               2
late_delivery                              2
traffic_level                              3
customer_type                              3
payment_method                             3
restaurant_type                            4
weather                                    4
city                                       4
discount_percent                           5
cancellation_reason                        5
restaurant_primary_category                6
day_of_week                                7
items_count                               10
delivery_partner_rating                   15
restaurant_rating                         21
order_hour                                24
order_date                                30
customer_rating                           33
customer_age                              37
delivery_area                             40
delivery_partner_experience_months        55
restaurant

## Data Quality Analysis

In [18]:
missing = df.isnull().sum().sort_values(ascending=False)
missing

cancellation_reason                    49132
customer_rating                          868
late_delivery                            868
actual_delivery_time_minutes             868
tip_amount                               868
weather                                    0
delivery_fee                               0
order_total                                0
payment_method                             0
distance_km                                0
delivery_partner_experience_months         0
traffic_level                              0
tax_amount                                 0
delivery_partner_rating                    0
restaurant_preparation_time_minutes        0
estimated_delivery_time_minutes            0
order_status                               0
service_fee                                0
order_id                                   0
customer_id                                0
subtotal                                   0
restaurant_id                              0
driver_id 

In [19]:
# missing Value in percentage 
missing_percentage = (
    df.isnull().sum() / len(df) * 100
).sort_values(ascending=False)

missing_percentage

cancellation_reason                    98.264
customer_rating                         1.736
late_delivery                           1.736
actual_delivery_time_minutes            1.736
tip_amount                              1.736
weather                                 0.000
delivery_fee                            0.000
order_total                             0.000
payment_method                          0.000
distance_km                             0.000
delivery_partner_experience_months      0.000
traffic_level                           0.000
tax_amount                              0.000
delivery_partner_rating                 0.000
restaurant_preparation_time_minutes     0.000
estimated_delivery_time_minutes         0.000
order_status                            0.000
service_fee                             0.000
order_id                                0.000
customer_id                             0.000
subtotal                                0.000
restaurant_id                     

## Duplicate Analysis

In [20]:
df[df['order_status'] == 'Cancelled']['cancellation_reason'].isnull().sum()

np.int64(0)

In [21]:
df.duplicated().sum()

np.int64(0)

## Data type Cheking 

In [22]:
df.dtypes

order_id                                object
customer_id                             object
restaurant_id                           object
driver_id                               object
order_timestamp                         object
order_date                              object
order_hour                               int64
day_of_week                             object
is_weekend                               int64
city                                    object
delivery_area                           object
customer_age                             int64
customer_type                           object
restaurant_type                         object
restaurant_primary_category             object
restaurant_rating                      float64
items_count                              int64
subtotal                               float64
discount_percent                         int64
tax_amount                             float64
service_fee                            float64
delivery_fee 

## Converting order_timestamp and order_date Into date time

In [23]:
df['order_timestamp'] = pd.to_datetime(df['order_timestamp'])
df['order_date'] = pd.to_datetime(df['order_date'])

In [24]:
df[['order_timestamp', 'order_date']].dtypes

order_timestamp    datetime64[ns]
order_date         datetime64[ns]
dtype: object


## Missing Values Handling

During the data cleaning process, missing values were checked using:

```python
df.isnull().sum()
```

The dataset contains missing values in a few columns. These missing values were not blindly filled because they have a valid reason.

### Missing Values Found

| Column                         | Missing Values | Reason                                                        |
| ------------------------------ | -------------: | ------------------------------------------------------------- |
| `cancellation_reason`          |         49,132 | Cancellation reason is not applicable to non-cancelled orders |
| `customer_rating`              |            868 | These orders were cancelled                                   |
| `late_delivery`                |            868 | Cancelled orders were not delivered                           |
| `actual_delivery_time_minutes` |            868 | No actual delivery time exists for cancelled orders           |
| `tip_amount`                   |            868 | No tip was recorded for cancelled orders                      |

### Handling Approach

* `cancellation_reason` → Kept as missing because it is not applicable to non-cancelled orders.
* `customer_rating` → Kept as missing because cancelled orders have no delivery rating.
* `late_delivery` → Kept as missing because cancelled orders were not delivered.
* `actual_delivery_time_minutes` → Kept as missing because delivery did not take place.
* `tip_amount` → Missing values can be replaced with `0` because no tip was recorded.

The 868 missing values in the delivery-related columns were confirmed to belong to **Cancelled orders**.

Therefore, these missing values are considered **meaningful missing values rather than data errors**.


## Delivery Delay -> Estimated vs actual delivery time

In [25]:
df['delivery_delay_minutes'] = (
    df['actual_delivery_time_minutes'] -
    df['estimated_delivery_time_minutes']
)

## Order Value Category -> Low / Medium / High order

In [26]:
df['order_value_category'] = pd.cut(
    df['order_total'],
    bins=[0, 300, 600, 1000, np.inf],
    labels=['Low', 'Medium', 'High', 'Very High']
)

## Distance Category

In [27]:
df['distance_category'] = pd.cut(
    df['distance_km'],
    bins=[0, 3, 7, 12, np.inf],
    labels=['Short', 'Medium', 'Long', 'Very Long']
)

## Age Group

In [28]:
df['age_group'] = pd.cut(
    df['customer_age'],
    bins=[0, 30, 45, 60, np.inf],
    labels=['18-30', '31-45', '46-60', '60+']
)

## Total Fees

In [29]:
df['total_fees'] = (
    df['tax_amount'] +
    df['service_fee'] +
    df['delivery_fee']
)

## Order Month

In [30]:
df['order_month'] = df['order_date'].dt.month

## Order Day

In [31]:
df['order_day'] = df['order_date'].dt.day_name()

## Is Late

In [32]:
df['is_late'] = df['delivery_delay_minutes'] > 0

## New Columns Data checking 

In [33]:
df[['delivery_delay_minutes',
    'order_value_category',
    'distance_category',
    'age_group',
    'total_fees',
    'order_month',
    'order_day',
    'is_late']].head()

,delivery_delay_minutes,order_value_category,distance_category,age_group,total_fees,order_month,order_day,is_late
0,4.0,Low,Short,31-45,11.47,10,Sunday,True
1,4.0,Low,Short,18-30,5.91,10,Wednesday,True
2,9.0,Low,Long,18-30,9.74,10,Tuesday,True
3,8.0,Low,Medium,18-30,6.22,10,Thursday,True
4,10.0,Low,Medium,18-30,7.39,10,Wednesday,True


## Exporting Cleaned Data CSV

After completing the data cleaning process and
creating the required new columns, 
the cleaned dataset was exported as a separate CSV file.



In [35]:
df.to_csv(r"C:\Users\rampa\DS Python Class\Food-Delivery-EDA\Data\cleaned/food_delivery_cleaned.csv", index=False)

## Overall Project Overview

In [34]:
print("Total Orders:", df['order_id'].nunique())
print("Total Customers:", df['customer_id'].nunique())
print("Total Restaurants:", df['restaurant_id'].nunique())
print("Total Revenue:", df['order_total'].sum())
print("Average Order Value:", df['order_total'].mean())

Total Orders: 50000
Total Customers: 4979
Total Restaurants: 250
Total Revenue: 1908475.05
Average Order Value: 38.169501000000004
